bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs

```diff
@@ -0,0 +1,270 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 Snowfork <hello@snowfork.com>
+//! Converts XCM messages into InboundMessage that can be processed by the Gateway contract
+
+use codec::DecodeAll;
+use core::slice::Iter;
+use frame_support::{ensure, BoundedVec};
+use snowbridge_core::{AgentIdOf, TokenId, TokenIdOf};
+
+use crate::v2::{
+	message::{Command, Message},
+	TransactInfo,
+};
+
+use sp_core::H160;
+use sp_runtime::traits::MaybeEquivalence;
+use sp_std::{iter::Peekable, marker::PhantomData, prelude::*};
+use xcm::prelude::*;
+use xcm_executor::traits::ConvertLocation;
+
+/// Errors that can be thrown to the pattern matching step.
+#[derive(PartialEq, Debug)]
+pub enum XcmConverterError {
+	UnexpectedEndOfXcm,
+	EndOfXcmMessageExpected,
+	WithdrawAssetExpected,
+	DepositAssetExpected,
+	NoReserveAssets,
+	FilterDoesNotConsumeAllAssets,
+	TooManyAssets,
+	ZeroAssetTransfer,
+	BeneficiaryResolutionFailed,
+	AssetResolutionFailed,
+	InvalidFeeAsset,
+	SetTopicExpected,
+	ReserveAssetDepositedExpected,
+	InvalidAsset,
+	UnexpectedInstruction,
+	TooManyCommands,
+	AliasOriginExpected,
+	InvalidOrigin,
+	TransactDecodeFailed,
+	TransactParamsDecodeFailed,
+	FeeAssetResolutionFailed,
+	CallContractValueInsufficient,
+}
+
+macro_rules! match_expression {
+	($expression:expr, $(|)? $( $pattern:pat_param )|+ $( if $guard: expr )?, $value:expr $(,)?) => {
+		match $expression {
+			$( $pattern )|+ $( if $guard )? => Some($value),
+			_ => None,
+		}
+	};
+}
+
+pub struct XcmConverter<'a, ConvertAssetId, Call> {
+	iter: Peekable<Iter<'a, Instruction<Call>>>,
+	ethereum_network: NetworkId,
+	_marker: PhantomData<ConvertAssetId>,
+}
+impl<'a, ConvertAssetId, Call> XcmConverter<'a, ConvertAssetId, Call>
+where
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+{
+	pub fn new(message: &'a Xcm<Call>, ethereum_network: NetworkId) -> Self {
```

Comment:
@yrong I know you didn't originally write this code, but the converter does not need to be instantiable. It does not hold any state. Should just be able to call a hypothetical static function `XcmConverter::convert(xcm)`.

I think if you make this change, it will no longer be necessary to pass the `'a` lifetime parameters.

See the inbound-queue message converter as an example of a static type that does conversion

Reply:
> It does not hold any state

It does hold the iter https://github.com/vgeddes/polkadot-sdk/blob/19f7f03635861fc31c6f5a9b2123585e44a2153a/bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs#L80 for iterating the instructions easily.

Alternatively, to make it stateless, we might need to pass the current instruction index into each function, though it is unclear if that would be more elegant.